platform/SharedPtr.h

```diff
@@ -0,0 +1,215 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __SHAREDPTR_H__
+#define __SHAREDPTR_H__
+
+#include <stdlib.h>
+
+#include <stdint.h>
+#include <stddef.h>
+
+/** Shared pointer class.
+  *
+  * Similar to std::shared_ptr in C++11.
+  *
+  * Usage: SharedPtr<class> POINTER(new class())
+  *
+  * When POINTER is passed around by value the copy constructor and
+  * destructor counts the number of references to the original object.
+  * If the counter reaches zero, delete is called on the object pointed to.
+  *
+  * To avoid loops, "weak" references should be used by calling the original
+  * pointer directly through POINTER.get().
+  */
+
+template <class T>
+class SharedPtr {
+public:
+    /**
+     * @brief Create empty SharedPtr not pointing to anything.
+     * @details Used for variable declaration.
+     */
+    SharedPtr(): pointer(NULL), counter(NULL) {
+    }
+
+    /**
+     * @brief Create new SharedPtr
+     * @param _pointer Pointer to take control over
+     */
+    SharedPtr(T* _pointer): pointer(_pointer) {
+        // allocate counter on the heap so it can be shared
+        counter = (uint32_t*) malloc(sizeof(uint32_t));
+        *counter = 1;
+    }
+
+    /**
+     * @brief Destructor.
+     * @details Decrement reference counter and delete object if no longer pointed to.
+     */
+    ~SharedPtr() {
+        decrementCounter();
+    }
+
+    /**
+     * @brief Copy constructor.
+     * @details Create new SharePointer from other SharedPtr by
+     *          copying pointer to original object and pointer to counter.
+     * @param source Object being copied from.
+     */
+    SharedPtr(const SharedPtr& source): pointer(source.pointer), counter(source.counter) {
+        // increment reference counter
+        if (counter) {
+            (*counter)++;
+        }
+    }
+
+    /**
+     * @brief Assignment operator.
+     * @details Cleanup previous reference and assign new pointer and counter.
+     * @param source Object being assigned from.
+     * @return Object being assigned.
+     */
+    SharedPtr operator=(const SharedPtr& source) {
+        if (this != &source) {
+            // clean up by decrementing counter
+            decrementCounter();
+
+            // assign new values
+            pointer = source.get();
+            counter = source.getCounter();
+
+            // increment new counter
+            if (counter) {
+                (*counter)++;
```

Comment:
Could you use an atomic so we get the same guarantee as the regular shared_ptr ?

Reply:
Hadn't realised it wasn't done atomically, will update the implementation :)